### PR TITLE
#9629 SelectOneMenu: filter does not work for lowercase i character

### LINKED ISSUE #9629: SelectOneMenu: filter does not work for accented characters
### Describe the bug

Problem is turkish "i" character  when search in items. selectOneMenu does not filtered with lowercase characters. but problem is on "i" character only. not on "ü", "ş" etc. other turkish characters work. if word is uppercase "DESTİNATİON", and if i search "DESTİ", its okay. but if i  search "desti", there is no result in items.  same on reserve scenerio. if word is lowercase like "destination"  and try to search "DESTİ". it does not work. this problem is at selectOneMenu only. example no problem at dataTable.

[problem is here](https://ibb.co/dtrTgwD)

### Reproducer

[primefaces-test.zip](https://github.com/primefaces/primefaces/files/10385219/primefaces-test.zip)

open selectOneMenu and search "edirn"

### Expected behavior

_No response_

### PrimeFaces edition

Community

### PrimeFaces version

12.0.0

### Theme

Serenity

### JSF implementation

Mojarra

### JSF version

2.2.11

### Java version

1.8

### Browser(s)

_No response_

### PR DESCRIPTION
**Pull request by jepsar**
Fix #9629

If this is okay, we can add the normalize function to other components with search filters.

Can be tested at http://localhost:8080/showcase/ui/input/oneMenu.xhtml and search at "Advanced" on "curac" to find "Curaçao".

### CODE REVIEW

**Review by melloware — APPROVED** [review]
This looks fantastic!   Did you check if we have any other places we might need this?

**Comment by melloware** [comment]
Damn build is getting `JSC_UNTRANSPILABLE. Cannot convert ECMASCRIPT_2018 feature "RegExp unicode property escape" to targeted output language. `   Althought now that we have dropped IE support I wonder if we can bump up the transpiler?

here is the browsers supporting it: https://caniuse.com/mdn-javascript_builtins_regexp_property_escapes

**Comment by jepsar** [comment]
> Damn build is getting `JSC_UNTRANSPILABLE. Cannot convert ECMASCRIPT_2018 feature "RegExp unicode property escape" to targeted output language. ` Althought now that we have dropped IE support I wonder if we can bump up the transpiler?
> 
> here is the browsers supporting it: https://caniuse.com/mdn-javascript_builtins_regexp_property_escapes

Was just looking at it. I refactored it.

**Comment by jepsar** [comment]
> This looks fantastic! Did you check if we have any other places we might need this?

I think these components:

* SelectOneListbox
* SelectManyMenu
* SplitButton
* PickList
* Tree
* TreeTable
* DataTable

**Review by melloware — APPROVED** [review]

File: primefaces/src/main/resources/META-INF/resources/primefaces/core/core.js [hunk 1]
```diff
@@ -1169,6 +1169,18 @@
...
         },
 
+        /**
+         * Normalizes the provided string.
+         * 
+         * @param {string} string to normalize.
+         * @param {boolean} lowercase flag indicating whether the string should be lower cased.
+         * @returns {string} to normalize.
+         */
+        normalize: function(string, lowercase) {
```
Comment (melloware) [inline primefaces/src/main/resources/META-INF/resources/primefaces/core/core.js:1179]: `if (!string) return string;`

**Review by melloware — CHANGES_REQUESTED** [review]

**Comment by melloware** [comment]
On more change I think since it a util function you should defend against NPE of someone passing in `undefined` or NULL string

File: primefaces/src/main/resources/META-INF/resources/primefaces/forms/forms.selectonemenu.js [hunk 2]
```diff
@@ -1288,7 +1288,7 @@ PrimeFaces.widget.SelectOneMenu = PrimeFaces.widget.DeferredWidget.extend({
 
             for(var i = 0; i < this.options.length; i++) {
                 var option = this.options.eq(i),
-                itemLabel = this.cfg.caseSensitive ? option.text() : option.text().toLowerCase(),
+                itemLabel = PrimeFaces.normalize(option.text(), !this.cfg.caseSensitive)
```
Comment (melloware) [inline primefaces/src/main/resources/META-INF/resources/primefaces/forms/forms.selectonemenu.js]: missing comma after this line?

**Comment by melloware** [comment]
Sorry if I stole your thunder too I answered on Stack Overflow even though that question was 7 years old.

**Comment by jepsar** [comment]
> Sorry if I stole your thunder too I answered on Stack Overflow even though that question was 7 years old.

No problem! Upvoted.

**Review by melloware — APPROVED** [review]

**Comment by jepsar** [comment]
> > This looks fantastic! Did you check if we have any other places we might need this?
> 
> I think these components:
> 
> * SelectOneListbox
> * SelectManyMenu
> * SplitButton
> * PickList
> * Tree
> * TreeTable
> * DataTable

I'll go over these components this evening to see if / how we can add normalization there.

**Pull request merged**